pkg/resource/deploy/source_eval.go

```diff
@@ -2197,7 +2227,28 @@ func (rm *resmon) RegisterResource(ctx context.Context,
 			Options: options,
 		})
 		if err != nil {
-			return nil, err
+			if st, ok := status.FromError(err); ok {
+				// We only turn errors into diagnostics for a few specific status codes that we know provider
+				// constructs return, and that aren't handled by the SDK.  The SDKs treat some status codes
+				// specially, and we want to keep that special handling, so we need to return the error verbatim
+				// to the SDK.
+				//nolint:exhaustive // Remaining cases are covered by default case.
+				switch st.Code() {
+				case codes.InvalidArgument, codes.NotFound, codes.AlreadyExists, codes.FailedPrecondition, codes.Aborted,
+					codes.OutOfRange, codes.DataLoss:
+					message := statusToMessage(st)
+					rm.diagnostics.Errorf(diag.GetResourceInvalidError(constructResult.URN), t, name, message)
+
+				default:
+					return nil, err
```

Comment:
I'm not sure it's good to have a difference in error handling between builtin errors and custom ones. If we exit the deployment due to failed construct we should _always_ exit on failed construct, not just on custom errors.

Reply:
Makes sense.  We do need to return the Unavailable/Canceled errors to the SDK, since it's specially handled at least in some SDKs, but I think all other errors we can get the status message, and abort early.

This also solves the case where we just throw whatever exception in the SDK, as that will come back as error code UNKNOWN, with a message, so it will be handled in the default case below.